lib/lev/status.rb

```diff
@@ -2,93 +2,142 @@
     STATE_FAILED = 'failed'
     STATE_KILLED = 'killed'
+    STATE_UNKNOWN = 'unknown'
 
     STATES = [
       STATE_QUEUED,
       STATE_WORKING,
       STATE_COMPLETED,
       STATE_FAILED,
-      STATE_KILLED
+      STATE_KILLED,
+      STATE_UNKNOWN
     ].freeze
 
-    attr_reader :uuid
+    def initialize(attrs = {})
+      @id = attrs[:id] || attrs['id'] || SecureRandom.uuid
+      @status = attrs[:status] || attrs['status'] || STATE_QUEUED
+      @progress = attrs[:progress] || attrs['progress'] || set_progress(0)
+      @errors = attrs[:errors] || attrs['errors'] || []
 
-    def initialize(uuid = nil)
-      @uuid = uuid || SecureRandom.uuid
-      save
+      set({ id: id,
+            status: status,
+            progress: progress,
+            errors: errors })
     end
 
-    def self.find(uuid)
-      if status = store.fetch(status_key(uuid))
-        JSON.parse(status)
+    def self.find(id)
+      attrs = { id: id }
+
+      if status = store.fetch(status_key(id))
+        attrs.merge!(JSON.parse(status))
       else
-        nil
+        attrs.merge!(status: STATE_UNKNOWN)
       end
+
+      new(attrs)
+    end
+
+    def self.all
+      job_ids.map { |id| find(id) }
     end
 
     def set_progress(at, out_of = nil)
       progress = compute_fractional_progress(at, out_of)
 
       data_to_set = { progress: progress }
-      data_to_set[:state] = STATE_COMPLETED if 1.0 == progress
+      data_to_set[:status] = STATE_COMPLETED if 1.0 == progress
 
       set(data_to_set)
+
+      progress
     end
 
     STATES.each do |state|
       define_method("#{state}!") do
-        set(state: state)
+        set(status: state)
       end
     end
 
-    def save(hash = {})
-      if has_reserved_keys?(hash)
-        raise IllegalArgument,
-              "Caller cannot specify any reserved keys (#{RESERVED_KEYS})"
+    def add_error(error, options = { })
+      @errors ||= []
+      options = { is_fatal: false }.merge(options)
+      @errors << { is_fatal: options[:is_fatal],
+                   code: error.code,
+                   message: error.message }
+      set(errors: @errors)
+    end
+
+    # Rails compatibility
+    # returns a Hash of all key-value pairs that have been #set()
+    def as_json(options = {})
+      stored
+    end
+
+    def save(incoming_hash)
+      if reserved = incoming_hash.select { |k, _| RESERVED_KEYS.include?(k) }.first
+        raise IllegalArgument, "Cannot set reserved key: #{reserved[0]}"
       else
-        set(hash)
+        set(incoming_hash)
       end
     end
 
-    def add_error(error, options = { })
-      options = { is_fatal: false }.merge(options)
-      push('errors', { is_fatal: options[:is_fatal],
-                       code: error.code,
-                       message: error.message })
+    def method_missing(method_name, *args)
+      instance_variable_get("@#{method_name}") || super
     end
 
-    def get(key)
-      self.class.find(uuid)[key]
+    def respond_to?(method_name)
```

Comment:
Line 99 is no go because method_name with a ? on the end isn't allowed as an instance variable name, and an exception is raised